File: FINAL_PROJECT/SudokuSolver_GUI/SudokuSolver_GUI/Gridset.vb
Imports Cell

Public Class Gridset

    Dim mCells = New Cell() {}

    Public Sub New()

    End Sub

    Public Sub AddCell(ByRef _Cell As Cell)
        Array.Resize(mCells, mCells.Length + 1)
        mCells(mCells.Length - 1) = _Cell
    End Sub

    Public Function IsValuePresent(_Value As Integer) As Boolean

        Dim cell As Cell
        For Each cell In mCells
            If cell.GetValue() = _Value Then
                Return True
            End If
        Next

        Return False

    End Function

    Public Function IsPossibleValuePresent(ByRef _Cell As Cell, _Value As Integer) As Boolean

        ' Walk through each Cell in the GridSet...
        Dim cell As Cell
        For Each cell In mCells
            If cell <> _Cell And cell.IsValuePossible(_Value) = True Then
                Return True
            End If
        Next

        Return False
    End Function

    Public Function FindUniquePossibleValue(ByRef _Cell As Cell, ByRef _Value As Integer) As Boolean

        ' Walk through each Cell in the GridSet...
        Dim cell As Cell
        For Each cell In mCells

            ' ...for each cell, determine if it contains a possible value that
            ' is unique within the GridSet...
            Dim uniqueValue As Integer = cell.FindUniquePossibleValueFromGridset(Me)

            If uniqueValue <> 0 Then

                ' FOUND!
                '
                ' Return the cell, and the value
                _Cell = cell
                _Value = uniqueValue

                Return True
            End If
        Next

        Return False

    End Function

    Public Function IsSolved() As Boolean

        Dim result As Boolean = True

        ' The GridSet is solved if all of the cells in the grid are populated
        ' with the positive integers from 1 to N, where N is the number of
        ' cells in the GridSet
        '
        ' We 'simply' (LOL) search for the presence of each integer, terminating if
        ' the first integer is not found.

        For searchValue As Integer = 1 To mCells.Length

            Dim cellResult As Boolean = False

            Dim cell As Cell
            For Each cell In mCells
                If cell.GetValue() = searchValue Then
                    cellResult = True
                    Exit For
                End If
            Next

            If cellResult = False Then
                result = False
            End If

            If result = False Then
                Exit For
            End If
        Next

        Return result
    End Function

End Class

File: FINAL_PROJECT/SudokuSolver_DLL/SudokuSolver_DLL/GridSet.cpp
//==============================================================================
// GridSet.cpp
//
// DESCRIPTION:
//   This class is a container for a set of cells. The set could represent
//   a row, a column, or a square of cells.
//
// &HISTORY.  
//   2006-08-28 PNM - Created.
//   2008-05-30 PNM - Updated comments.
// &EHISTORY.
//==============================================================================


//------------------------------------------------------------------------------
// HEADER FILES
//------------------------------------------------------------------------------
#include "stdafx.h"
#include <math.h>
#include "GridSet.h"


//------------------------------------------------------------------------------
// GridSet::GridSet()
//------------------------------------------------------------------------------
GridSet::GridSet()
{
}


//------------------------------------------------------------------------------
// GridSet::~GridSet()
//------------------------------------------------------------------------------
GridSet::~GridSet()
{
}


//------------------------------------------------------------------------------
// GridSet::AddCell()
//------------------------------------------------------------------------------
void GridSet::AddCell( Cell *_Cell )
{
  mCells.push_back( _Cell );
}


//------------------------------------------------------------------------------
// GridSet::IsValuePresent()
//------------------------------------------------------------------------------
bool GridSet::IsValuePresent( int _Value )
{
  for ( int index = 0 ; index < mCells.size() ; index++ )
    if ( mCells[ index ]->GetValue() == _Value )
      return( true );

  return( false );
}


//------------------------------------------------------------------------------
// GridSet::IsPossibleValuePresent()
//------------------------------------------------------------------------------
bool GridSet::IsPossibleValuePresent( Cell *_Cell, int _Value )
{
  //
  // Walk through each Cell in the GridSet...
  //
  for ( int index = 0 ; index < mCells.size() ; index++ )
  {
    // 
    // If the Cell is not the specified parameter cell, determine if
    // the specified value is possible.
    //
    if ( ( mCells[ index ] != _Cell ) &&
         ( mCells[ index ]->IsValuePossible( _Value ) ) )
      return( true );
  }

  return( false );  
}


//------------------------------------------------------------------------------
// GridSet::FindUniquePossibleValue()
//------------------------------------------------------------------------------
bool GridSet::FindUniquePossibleValue( Cell **_Cell, int *_Value )
{
  //
  // Walk through each Cell in the GridSet...
  //
  for ( int index = 0 ; index < mCells.size() ; index++ )
  {
    //
    // ...for each cell, determine if it contains a possible value that
    // is unique within the GridSet...
    //
    int UniqueValue =  mCells[ index ]->
      FindUniquePossibleValueFromGridSet( this );

    if ( UniqueValue != 0 )
    {
      // 
      // FOUND!
      //
      // Return the cell, and the value.
      //
      *_Cell = mCells[ index ];
      *_Value = UniqueValue;      

      return( true );
    }
  }

  return( false );
}


//------------------------------------------------------------------------------
// GridSet::IsSolved()
//------------------------------------------------------------------------------
bool GridSet::IsSolved()
{
  bool result = true;

  // 
  // The GridSet is solved if all of the cells in the grid are populated
  // with the positive integers from 1 to N, where N is the number of
  // cells in the GridSet.
  //
  // We simply search for the presence of each integer, terminating if
  // the first integer is not found.
  //
  for ( int searchValue = 1 ; 
        searchValue <= mCells.size() && result ;
        searchValue++ )
  {
    int index;
    for ( index = 0 ; index < mCells.size() ; index++ )
      if ( mCells[index]->GetValue() == searchValue )
        break;

    if ( index == mCells.size() )
      result = false;
  }

  return( result );
}


